### PR TITLE
Animations subsystem perf fixes.

### PR DESCRIPTION
**Pull request by jmacato**
Related to #2894

### CODE REVIEW

**Comment by kekekeks** [comment]
Are you sure that it won't break transition initialization from XAML? I think need a proper way of lazy initialization instead.

**Comment by jmacato** [comment]
@kekekeks All transitions from ControlCatalog are done on XAML and it works. But we need a better way of doing lazy init just as you said.

**Comment by jmacato** [comment]
@MarchingCube Can you do another object count pass on this? I don't have any profiling tools at hand right now.

**Comment by MarchingCube** [comment]
@jmacato Will take a look today, a bit busy with other stuff at the moment.

**Comment by MarchingCube** [comment]
After this change after going through all pages in ControlCatalog I only have 1 instance of `Transitions` alive.

I wonder what kind of API we want to provide here, since in UWP `Transitions` is lazy as well and you will never get null. And in Avalonia you can get null.

**Comment by jmacato** [comment]
@MarchingCube hmm nothing will consume the `Transitions` property of `Animatables` other than the animations subsystem nor would it make sense for other code to do so im not much worried about transitions being null 99% of the time honestly.  We need to fix `AvaloniaPropertyRegistry` to not call getters prematurely just as @kekekeks said.

File: src/Avalonia.Animation/Animatable.cs [hunk 1]
```diff
@@ -66,19 +66,18 @@ public Transitions Transitions
...
-            if (e.Priority != BindingPriority.Animation && Transitions != null && _previousTransitions != null)
-            {
-                var match = Transitions.FirstOrDefault(x => x.Property == e.Property);
+            if (_transitions is null || _previousTransitions is null || e.Priority == BindingPriority.Animation) return;
 
-                if (match != null)
-                {
-                    if (_previousTransitions.TryGetValue(e.Property, out var dispose))
-                        dispose.Dispose();
+            var match = Transitions.FirstOrDefault(x => x.Property == e.Property);
```
Comment (MarchingCube) [inline src/Avalonia.Animation/Animatable.cs]: This line is responsible for like 90% `Func` allocations when launching ControlCatalog and it gets worse when you start clicking around. We should "UnLINQ" it.

image

**Review by MarchingCube — CHANGES_REQUESTED** [review]
This is going in the right direction, my concern would in the API as after your changes control will have `null` transitions and this is different than UWP (which always returns something).

**Review by MarchingCube — DISMISSED** [review]
LGTM

File: tests/Avalonia.Animation.UnitTests/TransitionsTests.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ public void Check_Transitions_Interpolation_Negative_Bounds_Clamp()
             {
                 var border = new Border
                 {
-                    Transitions =
+                    Transitions =
```
Comment (MarchingCube) [inline tests/Avalonia.Animation.UnitTests/TransitionsTests.cs]: I guess changes here can be reverted?

File: src/Avalonia.Animation/Animatable.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ public IClock Clock
                 (o, v) => o.Transitions = v);
 
         private Transitions _transitions;
-
+        private bool _isTransitionsSet = false;
```
Comment (MarchingCube) [inline src/Avalonia.Animation/Animatable.cs]: Not needed.

**Review by MarchingCube — CHANGES_REQUESTED** [review]
Rerun my memory check in dotMemory. Only one `Transitions` object is allocated, and extra LINQ allocations are gone as well 👍

**Comment by yatli** [comment]
@MarchingCube this patch breaks my animation code. The workaround is to trigger the getter immediately after the transition is being set:

```C#
        this.Transitions = transitions;
        this.Transitions; // trigger
```

File: src/Avalonia.Animation/Animatable.cs [hunk 4]
```diff
@@ -66,18 +67,20 @@ public Transitions Transitions
         /// <param name="e">The event args.</param>
         protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs e)
         {
-            if (e.Priority != BindingPriority.Animation && Transitions != null && _previousTransitions != null)
-            {
-                var match = Transitions.FirstOrDefault(x => x.Property == e.Property);
+            if (_transitions is null || _previousTransitions is null || e.Priority == BindingPriority.Animation) return;
```
Comment (yatli) [inline src/Avalonia.Animation/Animatable.cs]: maybe:
`if ((_transitions is null &&_previousTransitions is null) || e.Priority == BindingPriority.Animation) return;`

**Pull request merged**